core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/NumberFunction.java

```diff
@@ -0,0 +1,224 @@
+        typedDelegate = new AggregateFunction<Long>() {
+          @Override
+          public void beginAggregate() {
+            startLong();
+          }
+
+          @Override
+          public void update(StructuredRecord record) {
+            Long val = record.get(fieldName);
+            if (val != null) {
+              updateLong(val);
+            }
+          }
+
+          @Override
+          public Long finishAggregate() {
+            return finishLong();
+          }
+
+          @Override
+          public Schema getOutputSchema() {
+            return isNullable ? Schema.nullableOf(Schema.of(Schema.Type.LONG)) : Schema.of(Schema.Type.LONG);
+          }
+        };
+        break;
+      case FLOAT:
+        typedDelegate = new AggregateFunction<Float>() {
+          @Override
+          public void beginAggregate() {
+            startFloat();
+          }
+
+          @Override
+          public void update(StructuredRecord record) {
+            Float val = record.get(fieldName);
+            if (val != null) {
+              updateFloat(val);
+            }
+          }
+
+          @Override
+          public Float finishAggregate() {
+            return finishFloat();
+          }
+
+          @Override
+          public Schema getOutputSchema() {
+            return isNullable ? Schema.nullableOf(Schema.of(Schema.Type.FLOAT)) : Schema.of(Schema.Type.FLOAT);
+          }
+        };
+        break;
+      case DOUBLE:
+        typedDelegate = new AggregateFunction<Double>() {
+          @Override
+          public void beginAggregate() {
+            startDouble();
+          }
+
+          @Override
+          public void update(StructuredRecord record) {
+            Double val = record.get(fieldName);
+            if (val != null) {
+              updateDouble(val);
+            }
+          }
+
+          @Override
+          public Double finishAggregate() {
+            return finishDouble();
+          }
+
+          @Override
+          public Schema getOutputSchema() {
+            return isNullable ? Schema.nullableOf(Schema.of(Schema.Type.DOUBLE)) : Schema.of(Schema.Type.DOUBLE);
+          }
+        };
+        break;
+      default:
+        throw new IllegalArgumentException(String.format("Field '%s' is of unsupported non-numeric type '%s'. ",
+                                                         fieldName, fieldType));
+    }
+  }
+
+  @Override
+  public void beginAggregate() {
+    typedDelegate.beginAggregate();
+  }
+
+  @Override
+  public void update(StructuredRecord record) {
+    typedDelegate.update(record);
+  }
+
+  @Override
+  public Number finishAggregate() {
+    return typedDelegate.finishAggregate();
+  }
+
+  @Override
+  public Schema getOutputSchema() {
+    return typedDelegate.getOutputSchema();
+  }
+
+  protected abstract void startInt();
+
+  protected abstract void startLong();
+
+  protected abstract void startFloat();
+
+  protected abstract void startDouble();
```

Comment:
would it be better to have `start(NumericType type)` where NumericType can be Enum and we can set/update value based on that ?